### DISCUSSION TITLE
will libvips make full use of multicore to accelerate encoding when compress is open in tiffsave

### DISCUSSION THREAD

**Discussion by zhusihan-python** [libvips/libvips#3229] [Q&A]
hello sir, when i run this demo code you provided, i found only 2 big core is working, the total usage is range from 10% -30%, how can i make my cpu work harder when compressing the jpeg, my cpu is 13600kf which have 6 Performance-core and 8 Efficient-core
what i could thought of is:
1、change mozjpeg to libjpeg-turbo
2、libvips distribute heavy work to use multicore better
```
/* compile with
 *
 *      g++ -g -Wall try336.cpp `pkg-config vips-cpp --cflags --libs`
 */

#include <vips/vips8>

using namespace vips;

int
main( int argc, char **argv )
{     
        if (VIPS_INIT (argv[0]))
                vips_error_exit (NULL);

        VImage image = VImage::new_from_file(argv[1], VImage::option()
                ->set("access", "sequential"));
  
        image.write_to_file(argv[2], VImage::option()
                ->set("tile-width", 256)
                ->set("tile-height", 256)
                ->set("tile", true)
                ->set("pyramid", true)
                ->set("bigtiff", true)
                ->set("compression", "jpeg")
                ->set("Q", 80));

        return 0;
}
```

**Comment by jcupitt** [comment]
Hello @zhusihan-python,

libvips master will use many cores for decoding tiled TIFF images, for example:

```
$ vips copy wtc.jpg x.tif[tile,compression=jpeg]
$ time vips avg x.tif
117.855110

real	0m0.328s
user	0m2.209s
sys	0m0.090s
```

You can see that real time was only 0.3s, but it used 2.2s of CPU time, so about a 7x speedup with parallel tile decompression.

However, tiff compression is single-threaded, for example:

```
$ time vips copy wtc.jpg x.tif[tile,compression=jpeg]

real	0m0.640s
user	0m0.928s
sys	0m0.147s
```

Here libvips is overlapping the decompression of the original JPG source and the creation of the tiff file (so real < user by 30%), but it is not parallising compression, which remains single threaded.

Parallel compression would be a useful thing to add, but no one has implemented it yet.

**Comment by zhusihan-python** [comment]
thanks for the reply, there is another question, i can use Vimage->thumbnail_image to get a thumbnail, how can i write the thumbnail_image  into the tifffile before i call image.write_to_file

**Comment by zhusihan-python** [comment]
replace mozjpeg to libjpeg-turbo, tiffsave compress is 4x faster, maybe it will be 10x faster if multicore is used in the future
rebuild the libvips in windows 10 WSL 2, docker desktop
`./build.sh --without-mozjpeg all`
![1671455151513](https://user-images.githubusercontent.com/23504604/208433066-173750ac-7d0d-4ed4-b732-4f3e3f962a4f.png)
